tests/test_domain_py.py

```diff
@@ -1840,3 +1841,77 @@ def test_short_literal_types(app):
             [desc_content, ()],
         )],
     ))
+
+
+def test_function_pep_695(app):
+    text = """.. py:function:: func[T: int, U: (int, str), *V, **P]"""
+    doctree = restructuredtext.parse(app, text)
+    assert_node(doctree, (
+        addnodes.index,
+        [desc, (
+            [desc_signature, (
+                [desc_name, 'func'],
+                [desc_tparameterlist, (
+                    [desc_parameter, (
+                        [desc_sig_name, 'T'],
+                        [desc_sig_punctuation, ':'],
+                        desc_sig_space,
+                        [desc_sig_name, ([pending_xref, 'int'])],
+                    )],
+                    [desc_parameter, (
+                        [desc_sig_name, 'U'],
+                        [desc_sig_punctuation, ':'],
+                        desc_sig_space,
+                        [desc_sig_punctuation, '('],
+                        [desc_sig_name, (
+                            [pending_xref, 'int'],
+                            [desc_sig_punctuation, ','],
+                            desc_sig_space,
+                            [pending_xref, 'str'],
+                        )],
+                        [desc_sig_punctuation, ')'],
+                    )],
+                    [desc_parameter, (
+                        [desc_sig_operator, '*'],
+                        [desc_sig_name, 'V'],
+                    )],
+                    [desc_parameter, (
+                        [desc_sig_operator, '**'],
+                        [desc_sig_name, 'P'],
+                    )],
+                )],
+                [desc_parameterlist, ()],
+            )],
+            [desc_content, ()],
+        )],
+    ))
+
+
+def test_class_def_pep_695(app):
+    # type checkers should reject this but it does not raise a compilation error
```

Comment:
Maybe add a test that use square brackets both in the type params and the bases, like `class Dict[KT, VT](Mapping[KT, VT])`.

Also, consider pre-emptively adding a test for PEP 696 syntax like
```
def f[T: str = str](): pass
```

I haven't tried it but it kind of looks like that might work already without further changes.

Reply:
I had a look at PEP 696 and it seems we would be able to write something like `foo[int](x, y)` if we have `def foo[T=str](x: T, y: T)`. But is there a plan where we could write something like `foo[T=int](x, y)` instead of `foo[int](x, y)`?

(I'll implement the tests next week after I'm done with personal work).

Reply:
PEP 696 doesn't do anything to function calls, only definition sites, and in any case in Sphinx we only need to care about mimicking the syntax for function definitions, right?

Reply:
Oh yes, the typed "call" is only for classes. So, I think the following will not be possible:

```python

class Foo[T1=int,  T2=str]: ...

# cannot 
x: Foo[T2=int] = Foo()  

# ok
x: Foo[int, int] 
```

If we could use keywords, I would expect that we can also use the positional only or keyword only markers in the type parameters list (currently, I parse the list and assume that the parameter's kind is either "normal", "*args"-like or "**kwargs" but I don't expect a "positional only" or "keyword only" marker, e.g. ``class Foo[T, *, U=int]:``).